pkg/hbone/dialer.go

```diff
@@ -0,0 +1,173 @@
+
+	"istio.io/istio/security/pkg/pki/util"
+	istiolog "istio.io/pkg/log"
+)
+
+var log = istiolog.RegisterScope("hbone", "", 0)
+
+type Config struct {
+	ProxyAddress string
+	Headers      http.Header
+	TLS          *tls.Config
+}
+
+type Dialer interface {
+	proxy.Dialer
+	proxy.ContextDialer
+}
+
+func NewDialer(cfg Config) Dialer {
+	return &dialer{cfg: cfg}
+}
+
+type dialer struct {
+	cfg Config
+}
+
+func (d dialer) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	if network != "tcp" {
+		return net.Dial(network, address)
+	}
+	// TODO: use context
+	c, s := net.Pipe()
+	go func() {
+		err := proxyTo(s, d.cfg, address)
+		log.Infof("Tunnel complete: %v", err)
+	}()
+	return c, nil
+}
+
+func (d dialer) Dial(network, address string) (c net.Conn, err error) {
+	return d.DialContext(context.Background(), network, address)
+}
+
+func proxyTo(conn io.ReadWriteCloser, req Config, address string) error {
+	defer conn.Close()
+	t0 := time.Now()
+
+	url := "http://" + req.ProxyAddress
+	if req.TLS != nil {
+		url = "https://" + req.ProxyAddress
+	}
+	// Setup a pipe. We could just pass `conn` to `http.NewRequest`, but this has a few issues:
+	// * Less visibility into i/o
+	// * http will call conn.Close, which will close before we want to (finished writing response).
+	pr, pw := io.Pipe()
+	r, err := http.NewRequest("CONNECT", url, pr)
+	if err != nil {
+		return fmt.Errorf("new request: %v", err)
+	}
+	r.Host = address
+
+	var transport *http2.Transport
+	if req.TLS != nil {
+		transport = &http2.Transport{
+			TLSClientConfig: req.TLS,
+		}
+	} else {
+		transport = &http2.Transport{
+			// For h2c
+			AllowHTTP: true,
+			DialTLS: func(network, addr string, cfg *tls.Config) (net.Conn, error) {
+				return net.Dial(network, addr)
+			},
+		}
+	}
+
+	// Initiate CONNECT.
+	log.Infof("initiate CONNECT to %v via %v", r.Host, url)
+
+	resp, err := transport.RoundTrip(r)
+	if err != nil {
+		return fmt.Errorf("round trip: %v", err)
+	}
+	defer resp.Body.Close()
+	var remoteID string
+	if resp.TLS != nil && len(resp.TLS.PeerCertificates) > 0 {
+		ids, _ := util.ExtractIDs(resp.TLS.PeerCertificates[0].Extensions)
+		if len(ids) > 0 {
+			remoteID = ids[0]
+		}
+	}
+	log.WithLabels("host", r.Host, "remote", remoteID).Info("CONNECT established")
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		// handle upstream (mtp server) --> downstream (app)
+		copyBuffered(conn, resp.Body, log.WithLabels("name", "body to conn"))
+		wg.Done()
+	}()
+	// Copy from conn into the pipe, which will then be sent as part of the request
+	// handle upstream (mtp server) <-- downstream (app)
+	copyBuffered(pw, conn, log.WithLabels("name", "conn to pipe"))
+
+	wg.Wait()
+	log.Info("stream closed in ", time.Since(t0))
+	return nil
+}
+
+// TLSDialWithDialer is an implementation of tls.DialWithDialer that accepts a generic Dialer
+func TLSDialWithDialer(dialer Dialer, network, addr string, config *tls.Config) (*tls.Conn, error) {
```

Comment:
Why do we need this ? Normal pattern is create Dialer with config and call DialContext/Dial

Reply:
I am not sure another way to add TLS to a Dialer?